dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java

```diff
@@ -0,0 +1,374 @@
+
+    }
+
+    /**
+     * Entry point for (Contentlets, Host, Persona, Templates, Folders, FileAsset)
+     * @param asset
+     * @param parent
+     * @return
+     */
+    @Override
+    public String generateDeterministicIdBestEffort(final Versionable asset,
+            final Treeable parent) {
+        return enabled() ? bestEffortDeterministicId(hash(deterministicIdSeed(asset, parent)))
+                : uuidSupplier.get();
+    }
+
+    /**
+     * Entry point for Content-Types
+     * @param contentType CT to generate the id for
+     * @param contentTypeVarName sometimes the var-name isn't set on the contentType
+     * which is immutable in such cases we need to rely on this supplier.
+     * @return generated deterministic id
+     */
+    @Override
+    public String generateDeterministicIdBestEffort(final ContentType contentType,
+            final Supplier<String> contentTypeVarName) {
+        return enabled() ? bestEffortDeterministicContentId(
+                hash(deterministicIdSeed(contentType, contentTypeVarName)))
+                : uuidSupplier.get();
+    }
+
+    /**
+     * Given a Language this will evaluate the code and country code if any then generate a sha256 and finally will hash it out into a long val
+     * @param lang
+     * @return
+     */
+    @Override
+    public long generateDeterministicIdBestEffort(final Language lang) {
+        //You never know.. better to be safe than sorry
+        if("US".equals(lang.getCountryCode()) &&  "en".equals(lang.getLanguageCode())){
+          return 1;
+        }
+        return enabled() ? bestEffortDeterministicLanguageId(
+                longHash(hash(deterministicIdSeed(lang)))) : nextLangId();
+    }
+
+    @CloseDBIfOpened
+    private long bestEffortDeterministicLanguageId(final long hash) {
+
+        long candidateId = hash;
+        for (int i = 1; i <= MAX_ATTEMPTS; i++) {
+            if (!isLanguageId(candidateId)) {
+                return candidateId;
+            }
+            candidateId = nextLangId();
+        }
+        Logger.warn(DeterministicIdentifierAPIImpl.class,String.format("Attempt to generate language id has failed instead a non deterministic id (%s) was returned.",candidateId));
+        return candidateId;
+    }
+
+    private synchronized long nextLangId(){
+        return System.currentTimeMillis();
+    }
+
+    /**
+     * Test if the calculated hash has already been used
+     * @param hash
+     * @return
+     */
+    private boolean isLanguageId(final long hash){
+        return new DotConnect()
+                .setSQL("select count(*) as test from language where id =?")
+                .addParam(hash)
+                .getInt("test")>0;
+    }
+
+    /**
+     * basically takes a string and hash it into a long number
+     * Used to create a deterministic long representation out of a sha256
+     * @param string
+     * @return
+     */
+    private long longHash(final String string) {
```

Comment:
nice

Reply:
not sure if this should be expose in other place, may be UUIDUtils

Reply:
I think it's fine here